public/app/features/explore/Logs/LogsTable.tsx

```diff
@@ -141,25 +166,93 @@ export const LogsTable: React.FunctionComponent<Props> = (props) => {
           },
         });
       });
+
+      // Create object of label filters to filter out any columns not selected by the user
+      let labelFilters: Record<string, true> = {};
+      Object.keys(columnsWithMeta)
+        .filter((key) => !columnsWithMeta[key].active)
+        .forEach((key) => {
+          labelFilters[key] = true;
+        });
+
+      // We could be getting fresh data
+      const uniqueLabels = new Set<string>();
+      const logFrameLabels = logsFrame?.getAttributesAsLabels();
+
+      // Populate the set with all labels from latest dataframe
+      logFrameLabels?.forEach((labels) => {
+        Object.keys(labels).forEach((label) => {
+          uniqueLabels.add(label);
+        });
+      });
+
+      // Check if there are labels in the data, that aren't yet in the labelFilters, and set them to be hidden by the transform
+      Object.keys(labelFilters).forEach((label) => {
+        if (!uniqueLabels.has(label)) {
+          labelFilters[label] = true;
+        }
+      });
+
+      // Check if there are labels in the label filters that aren't yet in the data, and set those to also be hidden
+      // The next time the column filters are synced any extras will be removed
+      Array.from(uniqueLabels).forEach((label) => {
+        if (label in columnsWithMeta && !columnsWithMeta[label]?.active) {
+          labelFilters[label] = true;
+        } else if (!labelFilters[label] && !(label in columnsWithMeta)) {
+          labelFilters[label] = true;
+        }
+      });
+
+      // Add one transform to remove everything we added above
+      if (Object.keys(labelFilters).length > 0) {
+        transformations.push({
+          id: 'organize',
+          options: {
+            excludeByName: labelFilters,
+          },
+        });
+      }
+
       if (transformations.length > 0) {
-        const [transformedDataFrame] = await lastValueFrom(transformDataFrame(transformations, [dataFrame]));
-        setTableFrame(prepareTableFrame(transformedDataFrame));
+        const transformedDataFrame = await lastValueFrom(transformDataFrame(transformations, [dataFrame]));
+        const tableFrame = prepareTableFrame(transformedDataFrame[0]);
+        setTableFrame(tableFrame);
       } else {
         setTableFrame(prepareTableFrame(dataFrame));
       }
     };
     prepare();
-  }, [prepareTableFrame, logsFrames, logsSortOrder]);
+
+    // This is messy: this is triggered either by user action of adding/removing columns, or from the dataframe (query result) changing
+    // We have to be careful and make sure not to assume that the columnsWithMeta is "recent", it is also updated when the dataframe changes, but it doesn't always happen first!
+    // Also if you add both the logFrame and the logFrameRaw as dependencies it will cause an infinite re-render: adding the eslint ignore
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
In general, if this is needed, it can be a sign that this effect should be refactored into something else.

Reply:
Agreed, this has been haunting me these past few days how to organize this better.

Reply:
So I guess the question is, should we release this to ops next week so we can get some internal feedback on the feature set (which gives us some time to clean it up), or should we hold off on release until we've had a chance to figure this out?

We can figure this out next week, I'd like to ask Sven a few questions before merge anyway, in the mean while I'll open another PR and see if I can rewrite this without breaking everything